request.go

```diff
@@ -0,0 +1,111 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tspclient
+
+import (
+	"crypto"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/notaryproject/tspclient-go/internal/hashutil"
+	"github.com/notaryproject/tspclient-go/internal/oid"
+)
+
+// MessageImprint contains the hash of the datum to be time-stamped.
+//
+//	MessageImprint ::= SEQUENCE {
+//	 hashAlgorithm   AlgorithmIdentifier,
+//	 hashedMessage   OCTET STRING }
+type MessageImprint struct {
+	HashAlgorithm pkix.AlgorithmIdentifier
+	HashedMessage []byte
+}
+
+// Request is a time-stamping request.
+//
+//	TimeStampReq ::= SEQUENCE {
+//	 version         INTEGER                 { v1(1) },
+//	 messageImprint  MessageImprint,
+//	 reqPolicy       TSAPolicyID              OPTIONAL,
+//	 nonce           INTEGER                  OPTIONAL,
+//	 certReq         BOOLEAN                  DEFAULT FALSE,
+//	 extensions      [0] IMPLICIT Extensions  OPTIONAL }
+type Request struct {
+	Version        int // fixed to 1 as defined in RFC 3161 2.4.1 Request Format
+	MessageImprint MessageImprint
+	ReqPolicy      asn1.ObjectIdentifier `asn1:"optional"`
+	Nonce          *big.Int              `asn1:"optional"`
+	CertReq        bool                  `asn1:"optional,default:false"`
+	Extensions     []pkix.Extension      `asn1:"optional,tag:0"`
+}
+
+// NewRequest creates a request sent to TSA based on the given digest and
+// hash algorithm.
+//
+// Parameters:
+// digest - hashedMessage of timestamping request's messageImprint
+// alg - hashing algorithm. Supported values are SHA256, SHA384, and SHA512
+func NewRequest(digest []byte, alg crypto.Hash) (*Request, error) {
+	hashAlg, err := oid.FromHash(alg)
+	if err != nil {
+		return nil, MalformedRequestError{Msg: err.Error()}
+	}
+	if len(digest) != alg.Size() {
+		return nil, MalformedRequestError{Msg: fmt.Sprintf("digest is of incorrect size: %d", len(digest))}
+	}
+	return &Request{
+		Version: 1,
+		MessageImprint: MessageImprint{
+			HashAlgorithm: pkix.AlgorithmIdentifier{
+				Algorithm: hashAlg,
+			},
+			HashedMessage: digest,
+		},
+		CertReq: true,
```

Comment:
We should populate nonce and check that in response.

Reply:
@priteshbandi `nonce` is actually OPTIONAL according to RFC 3161: https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.1.
Since we made the `Request` struct public, any user can build their own TSA request from there. The `NewRequest` is just a helper function with all the needs from Notation. (consider Notation as a user of tspclient-go).

Reply:
Code updated.

Reply:
none also provide added security protection against replay attack and it's recommended. https://datatracker.ietf.org/doc/html/rfc3161#section-4

Reply:
Code updated.